_includes/v19.1/app/insecure/gorm-sample.go

```diff
@@ -0,0 +1,158 @@
+	if err := transferFunds(db, fromAccount, toAccount, amount); err != nil {
+		// If the error is returned, it's either:
+		//   1. Not a transaction retry error, i.e., some other kind of
+		//   database error that you should handle here.
+		//   2. A transaction retry error that has occurred more than N
+		//   times (defined by the `maxRetries` variable inside
+		//   `transferFunds`), in which case you will need to figure out
+		//   why your database access is resulting in so much contention
+		//   (see 'Understanding and avoiding transaction contention':
+		//   https://www.cockroachlabs.com/docs/stable/performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention)
+		fmt.Println(err)
+	}
+
+	// Print balances after transfer to ensure that it worked.
+	printBalances(db)
+
+	// Delete accounts so we can start fresh when we want to run this
+	// program again.
+	deleteAccounts(db)
+}
+
+func transferFunds(db *gorm.DB, fromAccount Account, toAccount Account, amount int) error {
+	if fromAccount.Balance < amount {
+		return fmt.Errorf("account %d balance %d is lower than transfer amount %d", fromAccount.ID, fromAccount.Balance, amount)
+	}
+
+	var maxRetries = 3
+	for retries := 0; retries <= maxRetries; retries++ {
+		if retries == maxRetries {
+			return fmt.Errorf("hit max of %d retries, aborting", retries)
+		}
+
+		txn := db.Begin()
+		// The statement below is used for forcing transaction retries
+		// for verifying the retry loop logic.  It is not necessary
+		// for production code.
+		// txn.Exec("SELECT now()")
+		if err := txn.Exec(
+			// The below statement is used for forcing transaction
+			// retries for verifying the retry loop logic.  It is not
+			// necessary for production code.
+			// `SELECT crdb_internal.force_retry('1s'::INTERVAL)`,
+			`UPSERT INTO accounts (id, balance) VALUES
+			   (?, ((SELECT balance FROM accounts WHERE id = ?) - ?)),
+			   (?, ((SELECT balance FROM accounts WHERE id = ?) + ?))`,
+			fromAccount.ID, fromAccount.ID, amount,
+			toAccount.ID, toAccount.ID, amount,
+		).Error; err != nil {
+
+			// We need to cast GORM's db.Error to *pq.Error so we can
+			// detect the Postgres transaction retry error code and
+			// handle retries appropriately.
+			pqErr := err.(*pq.Error)
+
+			if pqErr.Code == "40001" {
+				// Since this is a transaction retry error, we
+				// ROLLBACK the transaction and sleep a little before
+				// trying again.  Each time through the loop we sleep
+				// for a little longer than the last time
+				// (A.K.A. exponential backoff).
+				txn.Rollback()
+				var sleepMs = math.Pow(2, float64(retries)) * 100 * (rand.Float64() + 0.5)
+				fmt.Printf("Hit 40001 transaction retry error, sleeping %s milliseconds\n", sleepMs)
+				time.Sleep(time.Millisecond * time.Duration(sleepMs))
+			} else {
+				return err
+			}
+		} else {
+			// Happy case.  All went well, so we commit and break out
+			// of the retry loop.
+			txn.Commit()
```

Comment:
`txn.Commit` can return an error which must be handled. In particular it can return a retry error.

Reply:
Updated to check for 40001 errors on `txn.Commit()` and continue the retry loop if they're encountered.